radicle-tracker/src/ops/id.rs

```diff
@@ -0,0 +1,99 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::metadata::clock::TimeDiff;
+use std::{cmp::Ordering, hash::Hash, time::SystemTime};
+use uuid::Uuid;
+
+pub trait Gen {
+    fn gen() -> Self;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Unique {
+    blob: Vec<u8>,
+}
+
+impl Unique {
+    pub fn new(id: &[u8]) -> Option<Self> {
+        if id.is_empty() {
+            return None;
+        }
+
+        Some(Unique { blob: id.to_vec() })
+    }
+
+    pub fn val(&self) -> &[u8] {
+        &self.blob
+    }
+}
+
+impl Gen for Unique {
+    fn gen() -> Self {
+        Unique {
+            blob: Uuid::new_v4().as_bytes().to_vec(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Timestamp {
+    time: TimeDiff,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct UniqueTimestamp {
```

Comment:
Something that I find annoying is that `v1` returns a `Result`: https://docs.rs/uuid/0.8.1/src/uuid/v1.rs.html#193-216. It looks like it does this if the bytes you pass in aren't of length 6 (so why didn't they just use `[u8; 6]` :thinking:)

But it does make sense since it's also a timestamp generated UUID plus some node id, which we could derive from some user information maybe?

Reply:
AFAIK v1 UUIDs are unique IFF "node ids" are unique, which means they depend on the uniqueness of the MAC address used when generating the UUID.
My problem with that is that we might want to support more than one device per network card (and the whole concept of "uniqueness of a network interface" is flaky anyway).

In our scenario it might work if we use device IDs as node ids but I did not to want to mess with it.
This is why I went for v4 in my attempt.

Reply:
The second argument to `UUID::new_v1` is (six bytes of) the node id, so.. :man_shrugging:

Reply:
I think I may be lost here. @kim Do you think using `v1` would be advantageous?